### PR TITLE
Implemented missing response codes.

### PR DESCRIPTION
**Pull request by Emm54321**
Implemented ALERT, BADCHARSET, CAPABILITY, and PARSE response codes. See section 7.1 of rfc 3501.

### CODE REVIEW

File: imap-proto/src/parser/rfc3501.rs [hunk 1]
```diff
@@ -97,6 +97,12 @@ named!(resp_text_code_alert<ResponseCode>, do_parse!(
     (ResponseCode::Alert)
 ));
 
+named!(resp_text_code_capability<ResponseCode>, do_parse!(
+    tag_s!("CAPABILITY") >>
+    capabilities: many1!(capability) >>
```
Comment (djc) [inline imap-proto/src/parser/rfc3501.rs]: It looks like technically this requires that the list of capabilities includes `IMAP4rev1`, maybe we should error out if it does not?

File: imap-proto/src/parser/rfc3501.rs [hunk 2]
```diff
@@ -727,5 +747,27 @@ mod tests {
...
+        match parse_response(b"* NO [BADCHARSET] error\r\n") {
+            Ok((_, Response::Data {
+                status: Status::No,
+                code: Some(ResponseCode::BadCharset(None)),
+                information: Some("error")
+            })) => {},
+            rsp @ _ => panic!("unexpected response {:?}", rsp)
+        }
+
+        match parse_response(b"* NO [BADCHARSET (utf-8 latin1)] error\r\n") {
```
Comment (djc) [inline imap-proto/src/parser/rfc3501.rs:795]: Would you mind adding a test here that has an empty list of charsets?

File: imap-proto/src/parser/rfc3501.rs (hunk 1 above)
Comment (Emm54321) [reply imap-proto/src/parser/rfc3501.rs]: I'm not sure such a test belongs to the parser. The client may want to try using a server that doesn't include IMAP4rev1 in its capabilities.

File: imap-proto/src/parser/rfc3501.rs (hunk 2 above)
Comment (Emm54321) [reply imap-proto/src/parser/rfc3501.rs:795]: Ok, I added it.

File: imap-proto/src/parser/rfc3501.rs (hunk 1 above)
Comment (djc) [reply imap-proto/src/parser/rfc3501.rs]: The formal syntax in the RFC requires that it's there, so I think that makes it a parser-level check, wouldn't you agree?

Comment (Emm54321) [reply imap-proto/src/parser/rfc3501.rs]: Indeed, I hadn't noticed it was part of the grammar.

**Comment by djc** [comment]
Thanks! I'll probably simplify the capabilities code a bit later on.
